### PR TITLE
feat: add 'WithLogWriter' for api 'RunWithOpts'

### PR DESCRIPTION
**Pull request by zong-zhe**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [x] N
- [ ] Y 

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

File: pkg/opt/opt.go [hunk 1]
```diff
@@ -26,13 +26,12 @@ type CompileOptions struct {
...
 // MergeOptions will merge the input options.
+// When combining multiple options, the 'WithLogWriter' should be set at the end.
 func MergeOptions(opts ...CompileOptions) CompileOptions {
 	var opt = DefaultCompileOptions()
 	for _, o := range opts {
 		opt.Merge(*o.Option)
-		if o.writer != nil {
-			opt.writer = o.writer
-		}
+		opt.writer = o.writer
```
Comment (Peefy) [inline pkg/opt/opt.go:34]: This line of code will lead to non idempotence of options, meaning that different options will affect the final option result due to different order settings. It is recommended to adopt builder mode here.

An implementation reference: https://github.com/helm/helm/blob/main/pkg/getter/getter.go#L53

File: pkg/opt/opt.go [hunk 2]
```diff
@@ -25,58 +25,41 @@ type CompileOptions struct {
...
+func WithNoSumCheck(is bool) Option {
+	return func(opt *CompileOptions) {
+		opt.noSumCheck = is
+	}
+}
+
+// WithLogWriter will set the log writer of the compiler.
+func WithLogWriter(writer io.Writer) Option {
+	return func(opt *CompileOptions) {
+		opt.writer = writer
```
Comment (Peefy) [inline pkg/opt/opt.go:61]: s/opt/opts

File: pkg/opt/opt.go [hunk 3]
```diff
@@ -25,58 +25,41 @@ type CompileOptions struct {
...
 }
 
 // WithNoSumCheck will set the 'no_sum_check' flag.
-func WithNoSumCheck(is bool) CompileOptions {
-	var opt = DefaultCompileOptions()
-	opt.noSumCheck = is
-	return *opt
+func WithNoSumCheck(is bool) Option {
+	return func(opt *CompileOptions) {
+		opt.noSumCheck = is
```
Comment (Peefy) [inline pkg/opt/opt.go:54]: s/opt/opts

File: pkg/opt/opt.go [hunk 4]
```diff
@@ -25,58 +25,41 @@ type CompileOptions struct {
...
 }
 
 // WithEntry will add an entry to the compiler.
-func WithVendor(isVendor bool) CompileOptions {
-	var opt = DefaultCompileOptions()
-	opt.isVendor = isVendor
-	return *opt
+func WithVendor(isVendor bool) Option {
+	return func(opt *CompileOptions) {
+		opt.isVendor = isVendor
```
Comment (Peefy) [inline pkg/opt/opt.go:47]: s/opt/opts

File: pkg/api/kpm_run_test.go [hunk 5]
```diff
@@ -214,3 +215,32 @@ func TestRunWithOptsAndNoSumCheck(t *testing.T) {
...
+	old := os.Stdout
+	r, w, _ := os.Pipe()
+	os.Stdout = w
+
+	pathMainK := filepath.Join(pkgPath, "main.k")
+
+	_, err := RunWithOpts(
+		opt.WithEntries([]string{pathMainK}),
+		opt.WithKclOption(kcl.WithWorkDir(pkgPath)),
+		opt.WithLogWriter(nil),
```
Comment (Peefy) [inline pkg/api/kpm_run_test.go:235]: It is best to add different option orders to test and verify idempotence.

Comment (zong-zhe) [reply pkg/api/kpm_run_test.go]: fixed

File: pkg/opt/opt.go (hunk 4 above)
Comment (zong-zhe) [reply pkg/opt/opt.go]: fixed

File: pkg/opt/opt.go (hunk 3 above)
Comment (zong-zhe) [reply pkg/opt/opt.go]: fixed

File: pkg/opt/opt.go (hunk 2 above)
Comment (zong-zhe) [reply pkg/opt/opt.go]: fixed

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**